pkg/vsphere/toolbox/hgfs/archive.go

```diff
@@ -0,0 +1,227 @@
+func (a *archive) Size() int64 {
+	return a.size
+}
+
+// Mode implementation of the os.FileInfo interface method.
+func (a *archive) Mode() os.FileMode {
+	return 0600
+}
+
+// ModTime implementation of the os.FileInfo interface method.
+func (a *archive) ModTime() time.Time {
+	return time.Now()
+}
+
+// IsDir implementation of the os.FileInfo interface method.
+func (a *archive) IsDir() bool {
+	return false
+}
+
+// Sys implementation of the os.FileInfo interface method.
+func (a *archive) Sys() interface{} {
+	return nil
+}
+
+// The trailer is required since TransferFromGuest requires a Content-Length,
+// which toolbox doesn't know ahead of time as the gzip'd tarball never touches the disk.
+// HTTP clients need to be aware of this and stop reading when they see the 2nd gzip header.
+var gzipHeader = []byte{0x1f, 0x8b, 0x08} // rfc1952 {ID1, ID2, CM}
+
+// newArchiveFromGuest returns an hgfs.File implementation to read a directory as a gzip'd tar.
+func newArchiveFromGuest(dir string) (File, error) {
+	r, w := io.Pipe()
+
+	a := &archive{
+		name:   dir,
+		Reader: r,
+		Writer: w,
+	}
+
+	gz := gzip.NewWriter(a.Writer)
+	tw := tar.NewWriter(gz)
+	a.done = r.Close
+
+	go func() {
+		err := a.pack(tw)
+
+		_ = tw.Close()
+		_ = gz.Close()
+		_, _ = w.Write(gzipHeader)
+		_ = w.CloseWithError(err)
+	}()
+
+	return a, nil
+}
+
+// newArchiveToGuest returns an hgfs.File implementation to expand a gzip'd tar into a directory.
+func newArchiveToGuest(dir string) (File, error) {
+	r, w := io.Pipe()
+
+	a := &archive{
+		name:   dir,
+		Reader: r,
+		Writer: w,
+	}
+
+	var cerr error
+	var wg sync.WaitGroup
+
+	a.done = func() error {
+		_ = w.Close()
+		// We need to wait for unpack to finish to complete its work
+		// and to propagate the error if any to Close.
+		wg.Wait()
+		return cerr
+	}
+
+	wg.Add(1)
+	go func() {
+		gz, err := gzip.NewReader(a.Reader)
+		if err != nil {
+			_ = r.CloseWithError(err)
+			return
+		}
+
+		tr := tar.NewReader(gz)
+
+		cerr = a.unpack(tr)
+		_ = gz.Close()
+		_ = r.CloseWithError(cerr)
+		wg.Done()
+	}()
+
+	return a, nil
+}
+
+// unpack writes the contents of the given tar.Reader to the a.name directory.
+func (a *archive) unpack(tr *tar.Reader) error {
+	for {
+		header, err := tr.Next()
+		if err != nil {
+			if err == io.EOF {
+				return nil
+			}
+			return err
+		}
+
+		name := filepath.Join(a.Name(), header.Name)
+		mode := os.FileMode(header.Mode)
+
+		switch header.Typeflag {
+		case tar.TypeDir:
+			err = os.MkdirAll(name, mode)
+		case tar.TypeReg:
+			_ = os.MkdirAll(filepath.Dir(name), 0700)
+
+			var f *os.File
+
+			f, err = os.OpenFile(name, os.O_CREATE|os.O_RDWR|os.O_TRUNC, mode)
+			if err == nil {
+				_, cerr := io.Copy(f, tr)
+				err = f.Close()
+				if cerr != nil {
+					err = cerr
+				}
+			}
+		case tar.TypeSymlink:
+			err = os.Symlink(header.Linkname, name)
+		}
+
+		// TODO: Uid/Gid may not be meaningful here without some mapping.
```

Comment:
Created issue #5550 as a start, we may want a separate issue specific to how id mapping would be implemented.